File: cpu_moe_schedule_optimization/planners/hot_wide_planner.py
"""Hot-wide template planner: a fast planner built from features of model-searched plans.

The near-optimal plans found by the model-objective LNS (``model_lns.py``) under the
probe-calibrated event model share three features, which this planner hard-codes:

1. shape: a few wide lanes (8/16/32 threads) that hold the hottest experts, and all
   remaining cores in narrow bulk lanes (4 threads);
2. load balance: lanes finish together once each width's isolated load is scaled by the
   event/isolated ratio the searched plans show for that width (``lane_scale``);
3. order: inside a lane the largest expert starts first and the rest follow in ascending
   route order.

For every template (wide multiset plus bulk lanes) that packs into the LLC domains, the
planner assigns experts by LPT on heterogeneous lanes (hot experts land on wide lanes when
that lowers the scaled load) and keeps the template with the smallest scaled maximum lane
load. No event simulation runs on the request path.
"""

from __future__ import annotations

from dataclasses import dataclass
from heapq import heapify, heappop, heappush
from itertools import combinations_with_replacement
from typing import Mapping, Sequence

try:
    from .model_lns import Lane, pack_lanes, planner_tasks
except ImportError:  # pragma: no cover - direct script import
    from model_lns import Lane, pack_lanes, planner_tasks

# Event makespan / isolated (windowed) load of lanes in the v10-searched, 2T-free plans of the
# 18 E1 workloads, scored with the v11 model (tmp/fast_planner_20260920, median per width).
DEFAULT_LANE_SCALE = {4: 1.119, 8: 1.065, 16: 1.045, 32: 1.032}


_UNSET = object()


def _begins(tasks) -> tuple[int, ...]:
    """Core offset of each lane, in the order the tasks introduce them."""
    seen: dict[tuple[int, int], None] = {}
    for _, _, core, threads, _ in tasks:
        seen.setdefault((int(core), int(threads)), None)
    return tuple(core for core, _ in seen)


def _lanes_from_tasks(tasks) -> tuple[Lane, ...]:
    groups: dict[tuple[int, int], list[tuple[int, int]]] = {}
    for expert, routes, core, threads, _ in tasks:
        groups.setdefault((int(core), int(threads)), []).append((int(expert), int(routes)))
    return tuple(Lane(threads, tuple(members)) for (_, threads), members in groups.items())


@dataclass(frozen=True)
class HotWidePlan:
    lanes: tuple[Lane, ...]
    begins: tuple[int, ...]
    shape: tuple[int, ...]
    score_ns: float
    templates: int

    def tasks(self):
        """IntervalPlanner task tuples (expert, routes, core_begin, threads, dependencies)."""
        return planner_tasks(self.lanes, self.begins)


class HotWidePlanner:
    def __init__(
        self,
        model,
        *,
        num_cores: int = 80,
        domain_cores: Sequence[int] = (40, 40),
        bulk_width: int = 4,
        wide_widths: Sequence[int] = (8, 16, 32),
        max_wide_lanes: int = 3,
        max_wide_cores: int = 48,
        lane_scale: Mapping[int, float] | None = None,
        window_policy=None,
        use_native: bool = True,
    ):
        if sum(domain_cores) != num_cores:
            raise ValueError("domain cores must cover num_cores")
        self.model = model
        self.num_cores = int(num_cores)
        self.domain_cores = tuple(int(c) for c in domain_cores)
        self.bulk_width = int(bulk_width)
        self.lane_scale = dict(DEFAULT_LANE_SCALE if lane_scale is None else lane_scale)
        self.window_policy = window_policy
        self.use_native = bool(use_native)
        self.wide_widths = tuple(sorted(wide_widths, reverse=True))
        self.max_wide_lanes = int(max_wide_lanes)
        self.max_wide_cores = int(max_wide_cores)
        self._native_planner = _UNSET
        self._cost: dict[tuple[int, int], float] = {}
        self._raw: dict[tuple[int, int], float] = {}
        self._last_shape: tuple[int, ...] | None = None
        shapes = []
        for count in range(0, max_wide_lanes + 1):
            for combo in combinations_with_replacement(sorted(wide_widths, reverse=True), count):
                if sum(combo) > max_wide_cores:
                    continue
                bulk = (self.num_cores - sum(combo)) // self.bulk_width
                shape = tuple(combo) + (self.bulk_width,) * bulk
                if pack_lanes([Lane(w, ()) for w in shape], self.domain_cores) is not None:
                    shapes.append(shape)
        if not shapes:
            raise ValueError("no template packs into the LLC domains")
        self.shapes = tuple(dict.fromkeys(shapes))

    def raw_cost(self, routes: int, width: int) -> float:
        """Isolated time scaled by the window table, before the per-width lane scale.

        The native planner applies the lane scale itself, so it takes these values; caching them
        here keeps both paths paying the same price for the model lookups.
        """
        key = (routes, width)
        value = self._raw.get(key)
        if value is None:
            scale = 1.0 if self.window_policy is None else float(self.window_policy.time_scale(routes, width))
            value = float(self.model.T_iso(routes, width)) * scale
            self._raw[key] = value
        return value

    def cost(self, routes: int, width: int) -> float:
        key = (routes, width)
        value = self._cost.get(key)
        if value is None:
            scale = 1.0 if self.window_policy is None else float(self.window_policy.time_scale(routes, width))
            value = float(self.model.T_iso(routes, width)) * scale * self.lane_scale.get(width, 1.0)
            self._cost[key] = value
        return value

    def assign(self, experts: Sequence[tuple[int, int]], shape: Sequence[int],
               costs: Mapping[int, Sequence[float]] | None = None,
               limit: float = float("inf")) -> tuple[tuple[Lane, ...], float] | tuple[None, float]:
        """LPT over heterogeneous lanes, ties by lane index; lanes come back in hot-first order.

        ``experts`` must be sorted by descending routes when ``costs`` is given (the cost rows are
        indexed by position); :meth:`plan` does that once for all templates. With ``limit`` the
        assignment stops as soon as a lane passes that load (LPT loads only grow), and returns
        ``(None, load)``: the template cannot beat the incumbent.
        """
        if costs is None:
            experts = sorted(experts, key=lambda item: (-item[1], item[0]))
            costs = {width: [self.cost(routes, width) for _, routes in experts] for width in set(shape)}
        heaps: dict[int, list[list[float]]] = {}
        members: list[list[tuple[int, int]]] = [[] for _ in shape]
        loads = [0.0] * len(shape)
        for lane, width in enumerate(shape):
            heaps.setdefault(width, []).append([0.0, lane])
        for heap in heaps.values():
            heapify(heap)
        for index, expert in enumerate(experts):
            best = None
            for width, heap in heaps.items():
                candidate = (heap[0][0] + costs[width][index], heap[0][1], width)
                if best is None or candidate < best:
                    best = candidate
            load, lane, width = best
            heappop(heaps[width])
            heappush(heaps[width], [load, lane])
            members[lane].append(expert)
            loads[lane] = load
            if load >= limit:
                return None, load
        lanes = tuple(Lane(width, tuple(group[:1] + group[1:][::-1])) for width, group in zip(shape, members))
        return lanes, max(loads, default=0.0)

    def _native(self):
        """The C++ port of this planner, or None when the extension does not carry it."""
        if self._native_planner is not _UNSET:
            return self._native_planner
        self._native_planner = None
        if self.use_native:
            try:
                from fused_cpp import _moe_C
            except Exception:  # pragma: no cover - no extension in this environment
                _moe_C = None
            factory = getattr(_moe_C, "NativeHotWidePlanner", None) if _moe_C is not None else None
            if factory is not None:
                self._native_planner = factory(
                    num_cores=self.num_cores, domain_cores=list(self.domain_cores), bulk_width=self.bulk_width,
                    wide_widths=list(self.wide_widths), max_wide_lanes=self.max_wide_lanes,
                    max_wide_cores=self.max_wide_cores,
                    lane_scale=[(int(w), float(v)) for w, v in sorted(self.lane_scale.items())],
                )
        return self._native_planner

    def plan_native(self, experts: Sequence[tuple[int, int]]) -> HotWidePlan | None:
        """The same plan from the native planner, or None when it is unavailable.

        The costs cross the boundary as plain doubles - one row per width, before the lane
        scale - so the native side holds no model state and the two paths score identically.
        """
        native = self._native()
        if native is None:
            return None
        active = [(int(e), int(r)) for e, r in experts if int(r) > 0]
        if not active:
            raise ValueError("at least one active expert is required")
        widths = sorted({w for shape in self.shapes for w in shape})
        rows = [[self.raw_cost(routes, width) for _, routes in active] for width in widths]
        result = native.plan([e for e, _ in active], [r for _, r in active], widths, rows)
        return HotWidePlan(lanes=_lanes_from_tasks(result["tasks"]), begins=_begins(result["tasks"]),
                           shape=tuple(result["shape"]), score_ns=float(result["score_ns"]),
                           templates=int(result["templates"]))

    def plan(self, experts: Sequence[tuple[int, int]]) -> HotWidePlan:
        experts = sorted(((int(e), int(r)) for e, r in experts if int(r) > 0), key=lambda item: (-item[1], item[0]))
        if not experts:
            raise ValueError("at least one active expert is required")
        costs = {width: [self.cost(routes, width) for _, routes in experts]
                 for width in {w for shape in self.shapes for w in shape}}
        best = None
        # The template that won the previous call first: it usually survives and then prunes the rest.
        order = self.shapes if self._last_shape is None else (
            (self._last_shape,) + tuple(s for s in self.shapes if s != self._last_shape))
        for shape in order:
            lanes, score = self.assign(experts, shape, costs, float("inf") if best is None else best[1])
            if lanes is not None and (best is None or score < best[1]):
                best = (lanes, score, shape)
        lanes, score, shape = best
        self._last_shape = shape
        used = tuple(lane for lane in lanes if lane.experts)
        begins = pack_lanes(used, self.domain_cores)
        assert begins is not None  # a subset of a packable template packs
        return HotWidePlan(lanes=used, begins=tuple(begins), shape=tuple(shape), score_ns=score,
                           templates=len(self.shapes))


__all__ = ["DEFAULT_LANE_SCALE", "HotWidePlan", "HotWidePlanner"]

File: csrc/moe_planner/hot_wide_planner.cpp
// SPDX-License-Identifier: Apache-2.0
#include "hot_wide_planner.h"

#include <algorithm>
#include <limits>
#include <numeric>
#include <queue>
#include <stdexcept>

namespace moe_planner {
namespace {

// First-fit decreasing by width into the domain with the most free cores, as pack_lanes does.
bool PackLanes(const std::vector<int>& widths, const std::vector<int>& domain_cores, std::vector<int>* begins) {
  std::vector<int> free = domain_cores;
  std::vector<int> cursor(domain_cores.size(), 0);
  int running = 0;
  for (size_t domain = 0; domain < domain_cores.size(); ++domain) {
    cursor[domain] = running;
    running += domain_cores[domain];
  }
  std::vector<int> order(widths.size());
  std::iota(order.begin(), order.end(), 0);
  std::stable_sort(order.begin(), order.end(), [&widths](int a, int b) {
    return widths[a] != widths[b] ? widths[a] > widths[b] : a < b;
  });
  begins->assign(widths.size(), 0);
  for (int index : order) {
    size_t domain = 0;
    for (size_t candidate = 1; candidate < free.size(); ++candidate) {
      if (free[candidate] > free[domain]) {
        domain = candidate;
      }
    }
    if (free[domain] < widths[index]) {
      return false;
    }
    (*begins)[index] = cursor[domain];
    cursor[domain] += widths[index];
    free[domain] -= widths[index];
  }
  return true;
}

struct LaneHeapEntry {
  double load;
  int lane;
  bool operator>(const LaneHeapEntry& other) const {
    return load != other.load ? load > other.load : lane > other.lane;
  }
};

}  // namespace

NativeHotWidePlanner::NativeHotWidePlanner(int num_cores, std::vector<int> domain_cores, int bulk_width,
                                           std::vector<int> wide_widths, int max_wide_lanes, int max_wide_cores,
                                           std::vector<std::pair<int, double>> lane_scale)
    : num_cores_(num_cores),
      domain_cores_(std::move(domain_cores)),
      bulk_width_(bulk_width),
      lane_scale_(std::move(lane_scale)) {
  int total = 0;
  for (int cores : domain_cores_) {
    total += cores;
  }
  if (total != num_cores_) {
    throw std::invalid_argument("domain cores must cover num_cores");
  }
  if (bulk_width_ <= 0) {
    throw std::invalid_argument("bulk width must be positive");
  }
  std::sort(wide_widths.begin(), wide_widths.end(), std::greater<int>());
  // Every multiset of at most max_wide_lanes wide lanes, the rest of the machine in bulk lanes.
  std::vector<std::vector<int>> combos{{}};
  for (int count = 1; count <= max_wide_lanes; ++count) {
    std::vector<std::vector<int>> next;
    for (const std::vector<int>& combo : combos) {
      if (static_cast<int>(combo.size()) != count - 1) {
        continue;
      }
      for (int width : wide_widths) {
        if (!combo.empty() && width > combo.back()) {
          continue;  // combinations with replacement, non-increasing
        }
        std::vector<int> extended = combo;
        extended.push_back(width);
        next.push_back(std::move(extended));
      }
    }
    combos.insert(combos.end(), next.begin(), next.end());
  }
  for (const std::vector<int>& combo : combos) {
    int wide = 0;
    for (int width : combo) {
      wide += width;
    }
    if (wide > max_wide_cores) {
      continue;
    }
    std::vector<int> shape = combo;
    for (int lane = 0; lane < (num_cores_ - wide) / bulk_width_; ++lane) {
      shape.push_back(bulk_width_);
    }
    std::vector<int> begins;
    if (!PackLanes(shape, domain_cores_, &begins)) {
      continue;
    }
    if (std::find(shapes_.begin(), shapes_.end(), shape) == shapes_.end()) {
      shapes_.push_back(std::move(shape));
    }
  }
  if (shapes_.empty()) {
    throw std::invalid_argument("no template packs into the LLC domains");
  }
}

NativeHotWidePlanner::Plan NativeHotWidePlanner::PlanExperts(
    const std::vector<int>& expert_ids, const std::vector<int>& routes, const std::vector<int>& cost_widths,
    const std::vector<std::vector<double>>& costs_by_width) const {
  if (expert_ids.size() != routes.size()) {
    throw std::invalid_argument("expert ids and routes must have the same length");
  }
  if (cost_widths.size() != costs_by_width.size()) {
    throw std::invalid_argument("cost widths and cost rows must have the same length");
  }
  // Largest routes first, ties by expert id, as the Python planner sorts.
  std::vector<int> order;
  order.reserve(expert_ids.size());
  for (size_t index = 0; index < expert_ids.size(); ++index) {
    if (routes[index] > 0) {
      order.push_back(static_cast<int>(index));
    }
  }
  if (order.empty()) {
    throw std::invalid_argument("at least one active expert is required");
  }
  std::sort(order.begin(), order.end(), [&](int a, int b) {
    return routes[a] != routes[b] ? routes[a] > routes[b] : expert_ids[a] < expert_ids[b];
  });

  auto scale_of = [this](int width) {
    for (const std::pair<int, double>& entry : lane_scale_) {
      if (entry.first == width) {
        return entry.second;
      }
    }
    return 1.0;
  };
  auto row_of = [&](int width) {
    for (size_t index = 0; index < cost_widths.size(); ++index) {
      if (cost_widths[index] == width) {
        return static_cast<int>(index);
      }
    }
    throw std::invalid_argument("missing cost row for a template width");
  };

  std::vector<int> best_members;      // lane index per expert, in `order` sequence
  std::vector<int> best_shape;
  double best_score = std::numeric_limits<double>::infinity();
  std::vector<int> members(order.size());

  std::vector<int> sequence(shapes_.size());
  std::iota(sequence.begin(), sequence.end(), 0);
  if (last_shape_ >= 0 && last_shape_ < static_cast<int>(shapes_.size())) {
    // The template that won the previous call first: it usually survives and then prunes the rest.
    sequence.erase(sequence.begin() + last_shape_);
    sequence.insert(sequence.begin(), last_shape_);
  }

  for (int shape_index : sequence) {
    const std::vector<int>& shape = shapes_[shape_index];
    std::vector<std::priority_queue<LaneHeapEntry, std::vector<LaneHeapEntry>, std::greater<LaneHeapEntry>>> heaps;
    std::vector<int> heap_width;
    std::vector<double> scaled;
    std::vector<int> rows;
    for (size_t lane = 0; lane < shape.size(); ++lane) {
      int width = shape[lane];
      size_t slot = 0;
      while (slot < heap_width.size() && heap_width[slot] != width) {
        ++slot;
      }
      if (slot == heap_width.size()) {
        heap_width.push_back(width);
        heaps.emplace_back();
        rows.push_back(row_of(width));
        scaled.push_back(scale_of(width));
      }
      heaps[slot].push(LaneHeapEntry{0.0, static_cast<int>(lane)});
    }
    double worst = 0.0;
    bool aborted = false;
    for (size_t position = 0; position < order.size(); ++position) {
      int expert = order[position];
      double best_load = std::numeric_limits<double>::infinity();
      int best_lane = -1;
      size_t best_slot = 0;
      for (size_t slot = 0; slot < heaps.size(); ++slot) {
        const LaneHeapEntry& top = heaps[slot].top();
        double load = top.load + costs_by_width[rows[slot]][expert] * scaled[slot];
        if (load < best_load || (load == best_load && top.lane < best_lane)) {
          best_load = load;
          best_lane = top.lane;
          best_slot = slot;
        }
      }
      heaps[best_slot].pop();
      heaps[best_slot].push(LaneHeapEntry{best_load, best_lane});
      members[position] = best_lane;
      worst = std::max(worst, best_load);
      if (best_load >= best_score) {  // LPT loads only grow: this template cannot win
        aborted = true;
        break;
      }
    }
    if (!aborted && worst < best_score) {
      best_score = worst;
      best_members = members;
      best_shape = shape;
      last_shape_ = shape_index;
    }
  }
  if (best_shape.empty()) {
    throw std::runtime_error("no template produced an assignment");
  }

  // Lane contents in the searched plans' order: the hottest expert first, the rest ascending.
  std::vector<std::vector<int>> lanes(best_shape.size());
  for (size_t position = 0; position < order.size(); ++position) {
    lanes[best_members[position]].push_back(order[position]);
  }
  std::vector<int> used_widths;
  std::vector<std::vector<int>> used_lanes;
  for (size_t lane = 0; lane < lanes.size(); ++lane) {
    if (lanes[lane].empty()) {
      continue;
    }
    std::vector<int> entries = lanes[lane];
    if (entries.size() > 1) {
      std::reverse(entries.begin() + 1, entries.end());
    }
    used_lanes.push_back(std::move(entries));
    used_widths.push_back(best_shape[lane]);
  }
  std::vector<int> begins;
  if (!PackLanes(used_widths, domain_cores_, &begins)) {
    throw std::runtime_error("a subset of a packable template failed to pack");
  }

  Plan plan;
  plan.shape = best_shape;
  plan.score_ns = best_score;
  plan.templates = static_cast<int>(shapes_.size());
  for (size_t lane = 0; lane < used_lanes.size(); ++lane) {
    int previous = -1;
    for (int index : used_lanes[lane]) {
      IntervalTask task;
      task.expert_id = expert_ids[index];
      task.routes = routes[index];
      task.core_begin = begins[lane];
      task.threads = used_widths[lane];
      if (previous >= 0) {
        task.dependencies.push_back(previous);
      }
      plan.tasks.push_back(std::move(task));
      previous = static_cast<int>(plan.tasks.size()) - 1;
    }
  }
  return plan;
}

}  // namespace moe_planner
